src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+﻿using System.Diagnostics.Contracts;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ProducerConsumerStream : Stream
+    {
+        private TaskCompletionSource<WriteBufferWrapper> _buffer;
+        private TaskCompletionSource<int> _dataAvail;
+        private WriteBufferWrapper _currentBuffer;
+        private bool _disposed;
+
+        public ProducerConsumerStream()
+        {
+            _buffer = new TaskCompletionSource<WriteBufferWrapper>();
+            _dataAvail = new TaskCompletionSource<int>();
+            _currentBuffer = WriteBufferWrapper.EmptyContainer;
+        }
+
+        public override void Flush()
+        {
+            _dataAvail.TrySetResult(0);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotImplementedException();
```

Comment:
Looking at other streams, it looks like I actually need to use `NotSupportedException`. According to [msdn](https://msdn.microsoft.com/en-us/library/system.notsupportedexception%28v=vs.110%29.aspx), this is the correct exception to use.